docs/design_proposals/design-2_new_distribution_class.md

```diff
@@ -0,0 +1,41 @@
+# Introduce a new class of execution time distribution
+
+Zhenwen Dai (2019-04-16)
+
+## Motivation
+
+The current type of distribution class in MXFusion is a "symbolic" notation for probabilistic model and variational posterior definition. When developing sophisticated inference algorithm, one often faces the need of creating some intermediate distribution objects that do not exist in model/posterior definition. In the implementation of current inference algorithms, those distributions are explicitly represented in terms of their parameters. This is not ideal for memory organization and code reusing. If there are distribution objects for execution, those intermediate distributions can be better organization in an object-oriented fashion. Many distribution computation functions can be implemented on top of those execution time distribution classes such as Kullback-Leiber divergence. A good example of such distribution classes is the distribution class in Tensorflow (tf.distrbution).
+
+## Proposed Changes
+
```

Comment:
Will you add a section here on how the proposed changes affect the existing Distribution classes? Also on how this will or could extend to encompass functions as well? We probably want to strive to keep the two Factors consistent when it comes to things like where runtime vs graphical code lives.